parlai/tasks/wizard_of_wikipedia/agents.py

```diff
@@ -0,0 +1,734 @@
+        episode_done = entry_idx == (self.len_episode(episode_idx) - 1)
+
+        wizard_first = ('Wizard' in d['dialog'][0]['speaker'])
+        idx = entry_idx * 2 if wizard_first else (entry_idx * 2) + 1
+
+        # first, get knowledge
+        apprentice_ret_passages = wizard_ret_passages = {}
+
+        if not wizard_first or idx != 0:
+            apprentice_entry = d['dialog'][idx - 1]
+            apprentice_ret_passages = apprentice_entry['retrieved_passages']
+        if idx - 2 >= 0:
+            wizard_prev_entry = d['dialog'][idx - 2]
+            wizard_ret_passages = wizard_prev_entry['retrieved_passages']
+
+        chosen_topic = d.get('chosen_topic', '')
+        chosen_topic_passages = d['chosen_topic_passage']
+        chosen_topic = d.get('chosen_topic', '')
+
+        knowledge_dict = {
+            chosen_topic: chosen_topic_passages,
+        }
+        for ret_passes in [apprentice_ret_passages, wizard_ret_passages]:
+            for passage in ret_passes:
+                for k, v in passage.items():
+                    if k not in knowledge_dict.keys():
+                        knowledge_dict[k] = v
+
+        # then, get text
+        if idx == 0:
+            # first message - only have the chosen topic
+            text = chosen_topic
+        elif idx == 1:
+            # first response - only have the first message
+            text = '{}\n{}'.format(chosen_topic, apprentice_entry['text'])
+        else:
+            text = ''
+            if self.label_type == 'chosen_sent':
+                # if chosen_sent, add wizard response to dialog history
+                text += '{}\n'.format(wizard_prev_entry['text'])
+            text += apprentice_entry['text']
+
+        # next, get label
+        wizard_entry = d['dialog'][idx]
+        if self.label_type == 'response':
+            labels = [wizard_entry['text']]
+        else:
+            title, sentence = _get_chosen_title_and_sent(wizard_entry,
+                                                         knowledge_dict)
+            labels = ['{} {}'.format(title, sentence)]
+
+        # finally, get label_candidates
+        label_cands = [TOKEN_NOCHOSEN + ' ' + TOKEN_NOCHOSEN]
+        knowledge_str = ''
+        for title, passage in knowledge_dict.items():
+            for p in passage:
+                if self.knowledge_separator:
+                    cand = '{} {} {}'.format(title, TOKEN_KNOWLEDGE, p)
+                else:
+                    cand = '{} {}'.format(title, p)
+                knowledge_str += cand + '\n'
+                label_cands.append(cand)
+        if self.label_type == 'response':
+            if 'train' in self.datatype:
+                label_cands = []
+            else:
+                label_cands = wizard_entry.get('candidate_responses', [])
+
+        action = {
+            'id': 'WizardDialogKnowledgeTeacher',
+            'text': text,
+            'labels': labels,
+            'chosen_topic': chosen_topic,
```

Comment:
Is this field actually being used anywhere for training right now?

Reply:
The learning models are not using it, but we did use it in the retrieval system and it's also in the first line of the dialogue. Note that the dataset is out but the full models are not out yet - we are working on getting them into the repository.

Reply:
When are the full models likely to be out?

Will the code reading the 'knowledge' key in the action dict also be coupled with this release and will there be framework-level support (such as changes to train_step and eval_step in TorchGeneratorAgent or in a new Agent) added for training and consuming separate knowledge/fact encoders?